lib/devise/controllers/store_location.rb

```diff
@@ -56,11 +56,12 @@ def stored_location_key_for(resource_or_scope)
       def extract_path_from_location(location)
         uri = parse_uri(location)
 
-        if uri 
+        if Devise.redirect_whitelist.include?(location)
```

Comment:
It could be something like:
```
if Devise.redirect_whitelist.any? { |domain| location.start_with?(domain) }
```

Reply:
Good point. My initial thought was based on a chat I had recently with @tegon about the same thing. They said they preferred to wait for Rails to implement that kind of logic (it was attempted in https://github.com/rails/rails/pull/34412 but reverted soon after). Having a Devise setting for redirect_whitelist would mean that if Rails implemented the feature, Devise would need to go trough a deprecation cycle before removing the option.

My feeling would be to wait for Rails to implement the url sanitization logic instead of adding a `redirect_whitelist` to Devise. But I'm no maintainer, so Ill leave them decide what's best for the gem.

Reply:
I don't necessarily agree with that.  I think that rails implementation would be nice, but I don't see the downside of including this feature in Devise. They appear to be complimentary to each other. 

This method is handling sanitization of what gets stored in the session for a redirect, and so I think there is value in being able to whitelist the sanitization. 

The rails change appears to also only allow you to specify a "single host" or that "any host" is okay to redirect to. So by allowing Devise to handle whitelisting what's stored for the redirect you can combine it with the rails validations for more granularity.

I also don't think it's difficult to go through a deprecation process if it's eventually not needed anymore although it would be inconvenient. I'm not the maintainer either though 🤷‍♂

Reply:
I've changed this back to being a strict whitelist that must match exactly since people could still initiate open redirects if we just matched the domain. Things like:
`?redirect=http://safe.com@evil.com`

Would redirect to `evil.com` so if we include whitelisting then it must be exact.